### PR TITLE
[AIRFLOW-2803] Fix all ESLint issues

### PR DESCRIPTION
**Pull request by tedmiston**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2803

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR runs eslint to resolve all issues in .js files and JS inside HTML files.

More specific details while this is WIP are listed in a second comment below.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

This PR is focused on linting existing code to be compatible with eslint recently added in [AIRFLOW-2783](https://issues.apache.org/jira/browse/AIRFLOW-2783).

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

Updated documentation for running eslint commands / linting JavaScript.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by tedmiston** [comment]
Here's the process I followed at a high level:

```
brew upgrade nvm
nvm install --lts
cd airflow/www_rbac
npm update npm
npm install
npm run lint:fix
npm run lint
...repeat last step...
```

Before:

```
$ npm run lint

✖ 1056 problems (1003 errors, 53 warnings)
  697 errors, 0 warnings potentially fixable with the `--fix` option.
```

After:

```
$ npm run lint

✖ 32 problems (32 errors, 0 warnings)
```

More info:

- I started at 1056 linter problems and am down to 32 now.  Most of the remaining ones are related to variable scoping or shadowing.  If that's all that's left, this shouldn't take too much more work.
- The auto-fix worked pretty well as a base.  It made some mistakes where Jinja and JS are used together which I corrected by hand.
- There are a handful of cases where eslint is trying to parse Jinja as JavaScript:
	- airflow/www_rbac/templates/airflow/dags.html:46
	- airflow/www_rbac/templates/airflow/gantt.html:5
	- airflow/www_rbac/templates/airflow/graph.html:9
	- airflow/www_rbac/templates/airflow/ti_log.html:35
	- airflow/www_rbac/templates/airflow/tree.html:3
- I tried using `/* eslint-disable */ ... /* eslint-enable */` for these, but no dice since the errors are parser errors as opposed to linter errors (https://github.com/eslint/eslint/issues/8419).  For the meantime, I've added these files to the .eslintignore, but I'd welcome a better solution if you know of one.  Do you know if it's possible to exclude just some lines of a file from being parsed?
- Side note: There's a weird error being printed when running `npm run lint` which says that eslint is exiting with a non-zero status; however this does not occur when running the same command directly (`./node_modules/.bin/eslint --ignore-path=.eslintignore --ext .js,.html .`).  Any suggestion here would be welcome as well.
- **Finally, this linting covers all JS in `www_rbac`, so there's a decent chance I've accidentally broken something.  If anyone has availability to help test, it would be greatly appreciated.**

/cc @verdan @bolkedebruin

**Comment by verdan** [comment]
Thank you @tedmiston 👍 
Did you look into the `eslint` plugin for Jinja templates? I believe there are a couple of plugins available that would make sure to read JS code within the Jinja templates, respecting the Jinja variables and template tags.

File: airflow/www_rbac/.eslintignore
Comment (verdan) [inline airflow/www_rbac/.eslintignore]: These are the files which includes all the custom javascript in the codebase, and which need to be lint properly. We should not include them in `.eslintignore` at all and try to fix the issue. (or look into the eslint jinja template plugins)

File: airflow/www_rbac/static/js/clock.js
Comment (verdan) [inline airflow/www_rbac/static/js/clock.js:32]: This is the es6 syntax, and we should make sure that the compiled result is in es5, as not all browsers support es6. (especially IE)

File: airflow/www_rbac/templates/airflow/circles.html
Comment (verdan) [inline airflow/www_rbac/templates/airflow/circles.html]: Why the `i` and `flip` are let and not `const` type variables?

File: airflow/www_rbac/templates/airflow/circles.html
Comment (verdan) [inline airflow/www_rbac/templates/airflow/circles.html]: this could be `let`

File: airflow/www_rbac/templates/airflow/dag.html
Comment (verdan) [inline airflow/www_rbac/templates/airflow/dag.html]: could this be `let` ?

File: airflow/www_rbac/templates/airflow/dag_code.html
Comment (verdan) [inline airflow/www_rbac/templates/airflow/dag_code.html]: Can we please try to look for `eslint` setting to follow the indentation of the parent jinja template tag? I think It should follow the same indent as the parent `<script>` tag.

**Review by verdan — CHANGES_REQUESTED** [review]

File: airflow/www_rbac/static/js/clock.js
Comment (tedmiston) [reply airflow/www_rbac/static/js/clock.js:32]: Good point.  What are your thoughts on bringing in a babel build step to transpile the ES6 to ES5 vs writing as ES5?  Either as part of this issue or separately a follow up.

**Comment by tedmiston** [comment]
@verdan Thank you for the review!  I wasn't aware of eslint for Jinja plugin. I'll definitely check that out as well as the other comments.

File: airflow/www_rbac/templates/airflow/circles.html
Comment (tedmiston) [reply airflow/www_rbac/templates/airflow/circles.html]: both get reassigned - `i` on 54 and `flip` on 103 and 113

File: airflow/www_rbac/templates/airflow/circles.html
Comment (tedmiston) [reply airflow/www_rbac/templates/airflow/circles.html]: Does it get reassigned?  I chose const since the function ends on the next line so I assumed it went out of scope without being changed.

File: airflow/www_rbac/templates/airflow/dag.html
Comment (tedmiston) [reply airflow/www_rbac/templates/airflow/dag.html]: Sure, it could be.  I preferred const over let anywhere the variable didn't get reassigned.  Is there an advantage to using let over const here?  Are you thinking build the URL up as multiple lines vs one long line?

File: airflow/www_rbac/static/js/clock.js
Comment (verdan) [reply airflow/www_rbac/static/js/clock.js:32]: Please note that most of the custom JS is written inline in .html files, and we are not yet considering that javascript in webpack, that means, we won't be able to transpile that javascript to ES5. (which is fine for now)
I am working on another issue to extract all inline JS from html files to separate .js files. 
https://issues.apache.org/jira/browse/AIRFLOW-2804

My suggestion would be to implement the ES6->ES5 tranpilation as part of this issue. And once this PR gets merged, we'll be able to extract all inline JS into separate .js files. 
We already have a JIRA issue for that: https://issues.apache.org/jira/browse/AIRFLOW-2730

**Comment by verdan** [comment]
@tedmiston can you please make sure:
- you squash your commits 
- your commit message adheres the [commit guidelines](https://github.com/apache/incubator-airflow/blob/master/.github/PULL_REQUEST_TEMPLATE.md#commits)

**Comment by tedmiston** [comment]
@verdan Sure!  Typically I keep atomic commits while I'm working so everyone can follow small changes instead of one big diff, then squash down to one commit at the end.  I updated the title to make it clear this is WIP.  Since you're doing most of the reviewing here, do you have a preference on squashing throughout working vs just squashing pre-merge?

I should have an update later today btw.

**Comment by ashb** [comment]
FWIW I too am in favour of atomic/fixup! commits that then get squashed pre merge.

**Comment by r39132** [comment]
@verdan once @tedmiston is done, please provide your +1 and notify some of the committers on this PR that the PR is ready for validation and merge. Thx for your help on reviewing this PR!

File: airflow/www_rbac/static/js/clock.js
Comment (tedmiston) [reply airflow/www_rbac/static/js/clock.js:32]: Sounds good.  I will stick with the ES5 for now for this PR.

File: airflow/www_rbac/templates/airflow/circles.html
Comment (ashb) [reply airflow/www_rbac/templates/airflow/circles.html]: My JS is rusty - do `let` and `const` work across all browsers?

Comment (tedmiston) [reply airflow/www_rbac/templates/airflow/circles.html]: I don't write tons of modern JS either, but yes for Chrome, Firefox, Safari, IE, Edge, etc.

https://caniuse.com/#feat=let

https://caniuse.com/#feat=const

**Comment by verdan** [comment]
@tedmiston please tag me once it is ready for the next review. I see you're still working on this PR. i.e., Jinja template tags, indentation and some commented out code.

File: airflow/www_rbac/templates/airflow/circles.html
Comment (verdan) [reply airflow/www_rbac/templates/airflow/circles.html]: Also, the plan is to extract all the inline JS into separate .js files and then we'd be able to compile all JS to make it work across all browsers.

**Comment by r39132** [comment]
Any updates on this PR?

**Comment by verdan** [comment]
@r39132 @tedmiston I am working on extracting all inline JS from html files in another PR i.e., https://github.com/verdan/incubator-airflow/pull/10
I would suggest to close this PR for now and once all extraction is done, we can get back to this. We won't care about the .html files (Jinja templates) anymore and the transpiling issues would be easily solved. Thoughts?

**Comment by tedmiston** [comment]
Hey all - thanks for bumping.  Sorry for any delay here — my work sprint has been quite busy this past week.

This PR is actually very nearly done.  I have a local WIP commit to be pushed up with a small bit of cleanup left IIRC.  Given that, my preference would be to wrap up and merge this PR since I've already done ~95% of the work, then rebase/update @verdan's extraction PR above (hopefully this doesn't cause much extra work there).  This is something I can get done very soon.

I'm also happy to help review/test/etc on the extraction PR as well.  Does this work for you @verdan ?

**Comment by verdan** [comment]
@tedmiston sounds good to me!! 👍

**Comment by verdan** [comment]
Do we have any updates on this PR?

**Comment by tedmiston** [comment]
@verdan Yes! I'll push up my new work on this today.

**Comment by tedmiston** [comment]
@verdan Please see the revised diff / new squashed commit, and review when you have a chance.

There were some opinionated decisions to make here as far as which warnings / errors to fix via refactoring vs which were more complicated and worthy of disabling via a linter comment... there's a decent amount of tech debt in the front end JS.  I started off doing more refactoring, but eventually saw the size of this diff growing and growing and switched to only refactoring what was absolutely required for right now.

Some notes:

- This diff overall now resolves something like ~1500–2500 linter errors in total.  In some places I have used disable linter calls if refactoring the code to pass felt cumbersome or like a bit much.  I think these few long tail errors could be cleaned up with refactoring post-PR.
- I have an example of refactoring building the long query string with template strings here - https://github.com/apache/incubator-airflow/pull/3656/commits/b65388a7b772fa7b40140d77ef25c45ad1c475c9.  Since the PR grew pretty huge already, I'm thinking of maybe pulling it out of this PR and refactoring the others requests separately if this style pattern feels like an improvement.  Rewriting this also made me realize we have a dormant bug where some query params aren't being encoded which is extra motivation to migrate this part into a separate PR and keep scope here focused on linting.
- The original [eslint-plugin-jinja](https://github.com/alexkuz/eslint-plugin-jinja) is not maintained and [doesn't](https://github.com/alexkuz/eslint-plugin-jinja/issues/3) [work](https://github.com/alexkuz/eslint-plugin-jinja/issues/4) with current version of ESLint, however, the [jupyterlab/eslint-plugin-jinja](https://github.com/jupyterlab/eslint-plugin-jinja) fork does, so I went with that one.
- Apparently there's an issue with `eslint --fix` when using the eslint-plugin-jinja plugin / some plugins where it just doesn't work but also doesn't given an error.  "Unfortunately, auto-fix is not supported on plugins with processors (i.e., plugins which transform files to be ready for linting in JS)." (https://github.com/eslint/eslint/issues/7456).
- My workaround for this was to temporarily comment out the plugin to run `--fix` which will fix some things but also break some things (which the plugin would prevent) inside Jinja, but since it can auto fix many errors, it was still helpful.  Then when done, uncomment the plugin, fix any auto errors by hand, and re-run lint.  [I'm considering adding this to our docs since it was unexpected.]
- I have 2 FIXMEs outstanding - 1 in graph.html, 1 in tree.html where the linter is convinced the lines should not end with a semicolon (check the error).  I haven't figured out how to solve these yet if you have  any advice.
- Aligned all outermost vertical indentation to match level of script tags.
- Ensured consistent indentation for script tags nested inside Jinja blocks.
- Ensured newlines surrounding every function definition.
- Added blank newline after Apache license text where missing for consistency.
- Added missing closing </div> tag in circles.html.
- Needs tested from the GUI.
- Also just realized I forgot to revert the use of ES6 syntax for anonymous functions, so I'll have to fix that.  Otherwise, I'm hoping this is very close to done.

**Comment by r39132** [comment]
@verdan pls review @tedmiston please rebase

**Comment by verdan** [comment]
👍 looks great!! I believe this one is ready to be finalized. nice work @tedmiston

**Comment by r39132** [comment]
@tedmiston Please resolve conflicts and squash your commits. @verdan thx for your feedback. It would be great if you could do one more pass after the rebase & squash! Are the steps you outline above (i.e. `npm run lint`) still valid? Thx for your work on this btw!

**Comment by r39132** [comment]
@tedmiston any updates?

**Comment by tedmiston** [comment]
@r39132 Yes, I've been working on the revisions discussed above and anticipate pushing up the next pass later today or tomorrow.

Edit: The `npm run lint` is still valid.  I've updated CONTRIBUTING.md locally for a note with a workaround on how to use `npm run lint:fix` with the Jinja plugin.

**Comment by jmcarp** [comment]
What's the status of this PR @tedmiston? I was poking at the js and noticed a bunch of linter errors. Would be great to get this merged.

**Comment by tedmiston** [comment]
Hey @jmcarp thanks for bumping this... it looks like I forgot to close it out.  I think it's done shy of a rebase but I may have minor local changes on my machine.  I've blocked time tonight to dig into it again.

**Comment by feng-tao** [comment]
@tedmiston , any update on this pr? Are you still planning on working on fixing the issue?

**Comment by tedmiston** [comment]
Hey all - sorry I dropped off on this one over the holidays.

Yes, I can bring this back in and see what's required.  It looks like I'll have some time in the evening this week to close it out.

**Comment by mik-laj** [comment]
Hi.
I created a related PR: 
https://github.com/apache/airflow/pull/4787
This also fixes linting of the javascript that is in the HTML code.

I also worked on fixing these linter errors. I have not found your PR before.
https://github.com/PolideaInternal/airflow/pull/75
Firstly, I wanted to introduce linting with a list of exclusions. In the next PRs I would gradually fix the code.  I was afraid that changes in the JS code are difficult to test. Therefore, they should be carefully reviewed. It is difficult to review all changes if they affect each line.

The next step would be to extract Javascript code completely into separate files and introduce tests for them.

**Pull request closed without merging**
